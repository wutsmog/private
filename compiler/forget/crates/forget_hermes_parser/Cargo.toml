[package]
name = "forget_hermes_parser"
version = "0.1.0"
authors.workspace = true
description.workspace = true
edition.workspace = true
homepage.workspace = true
keywords.workspace = true
license.workspace = true
repository.workspace = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
insta = { workspace = true }
forget_diagnostics = { workspace = true }
forget_estree = { workspace = true }
hermes = { workspace = true } 
juno_support = { workspace = true }
serde_json = { workspace = true }

[build-dependencies]
forget_estree_codegen = { workspace = true }