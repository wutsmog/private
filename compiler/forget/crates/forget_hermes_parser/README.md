# forget_hermes_parser

Wrapper around the Hermes parser that exposes parse results as a `forget_estree` AST.