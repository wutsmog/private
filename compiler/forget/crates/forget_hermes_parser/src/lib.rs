mod generated;
mod generated_extension;

use forget_diagnostics::Diagnostic;
use forget_estree::Program;
use generated_extension::{Context, FromHermes};
use hermes::parser::{HermesParser, ParserDialect, ParserFlags};
use hermes::utf::utf8_with_surrogates_to_string;
use juno_support::NullTerminatedBuf;

pub fn parse(source: &str, _file: &str) -> Result<Program, Vec<Diagnostic>> {
    let buf = NullTerminatedBuf::from_str_check(source);
    let result = HermesParser::parse(
        ParserFlags {
            dialect: ParserDialect::TypeScript,
            enable_jsx: true,
            store_doc_block: true,
            strict_mode: true,
        },
        &buf,
    );
    let mut cx = Context::new();
    if result.has_errors() {
        let error_messages = result.messages();
        return Err(error_messages
            .into_iter()
            .map(|diag| {
                let message = utf8_with_surrogates_to_string(diag.message.as_slice()).unwrap();
                Diagnostic::invalid_syntax(message, None)
            })
            .collect());
    }

    Ok(FromHermes::convert(&mut cx, result.root().unwrap()))
}
