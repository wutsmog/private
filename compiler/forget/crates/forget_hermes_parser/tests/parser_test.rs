use std::env;

use forget_hermes_parser::parse;
use insta::{assert_snapshot, glob};
use serde_json;

#[test]
fn fixtures() {
    glob!("fixtures/**.js", |path| {
        println!("fixture {}", path.to_str().unwrap());
        let input = std::fs::read_to_string(path).unwrap();
        let ast = parse(&input, path.to_str().unwrap()).unwrap();
        let output = serde_json::to_string_pretty(&ast).unwrap();
        let output = output.trim();
        assert_snapshot!(format!("Input:\n{input}\n\nOutput:\n{output}"));
    });
}
