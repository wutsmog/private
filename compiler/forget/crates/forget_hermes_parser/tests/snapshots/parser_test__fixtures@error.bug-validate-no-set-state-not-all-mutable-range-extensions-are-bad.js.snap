---
source: crates/forget_hermes_parser/tests/parser_test.rs
expression: "format!(\"Input:\\n{input}\\n\\nOutput:\\n{output}\")"
input_file: crates/forget_hermes_parser/tests/fixtures/error.bug-validate-no-set-state-not-all-mutable-range-extensions-are-bad.js
---
Input:
function Component(props) {
  const logEvent = useLogging(props.appId);
  const [currentStep, setCurrentStep] = useState(0);

  const onSubmit = (errorEvent) => {
    // 2. onSubmit inherits the mutable range of logEvent
    logEvent(errorEvent);
    // 3. this call then triggers the ValidateNoSetStateInRender check incorrectly, even though
    //    onSubmit is not called during render (although it _could_ be, if OtherComponent does so.
    //    but we can't tell without x-file analysis)
    setCurrentStep(1);
  };

  switch (currentStep) {
    case 0:
      return <OtherComponent data={{ foo: "bar" }} />;
    case 1:
      return <OtherComponent data={{ foo: "joe" }} onSubmit={onSubmit} />;
    default:
      // 1. logEvent's mutable range is extended to this instruction
      logEvent("Invalid step");
      return <OtherComponent data={null} />;
  }
}


Output:
{
  "type": "Program",
  "body": [
    {
      "type": "FunctionDeclaration",
      "id": {
        "type": "Identifier",
        "name": "Component",
        "typeAnnotation": null,
        "loc": null,
        "range": {
          "start": 0,
          "end": 1
        }
      },
      "params": [
        {
          "type": "Identifier",
          "name": "props",
          "typeAnnotation": null,
          "loc": null,
          "range": {
            "start": 0,
            "end": 1
          }
        }
      ],
      "body": {
        "type": "BlockStatement",
        "body": [
          {
            "type": "VariableDeclaration",
            "kind": "const",
            "declarations": [
              {
                "type": "VariableDeclarator",
                "id": {
                  "type": "Identifier",
                  "name": "logEvent",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 0,
                    "end": 1
                  }
                },
                "init": {
                  "type": "CallExpression",
                  "callee": {
                    "type": "Identifier",
                    "name": "useLogging",
                    "typeAnnotation": null,
                    "loc": null,
                    "range": {
                      "start": 0,
                      "end": 1
                    }
                  },
                  "arguments": [
                    {
                      "type": "MemberExpression",
                      "object": {
                        "type": "Identifier",
                        "name": "props",
                        "typeAnnotation": null,
                        "loc": null,
                        "range": {
                          "start": 0,
                          "end": 1
                        }
                      },
                      "property": {
                        "type": "Identifier",
                        "name": "appId",
                        "typeAnnotation": null,
                        "loc": null,
                        "range": {
                          "start": 0,
                          "end": 1
                        }
                      },
                      "computed": false,
                      "loc": null,
                      "range": {
                        "start": 0,
                        "end": 1
                      }
                    }
                  ],
                  "loc": null,
                  "range": {
                    "start": 0,
                    "end": 1
                  }
                },
                "loc": null,
                "range": {
                  "start": 0,
                  "end": 1
                }
              }
            ],
            "loc": null,
            "range": {
              "start": 0,
              "end": 1
            }
          },
          {
            "type": "VariableDeclaration",
            "kind": "const",
            "declarations": [
              {
                "type": "VariableDeclarator",
                "id": {
                  "type": "ArrayPattern",
                  "elements": [
                    {
                      "type": "Identifier",
                      "name": "currentStep",
                      "typeAnnotation": null,
                      "loc": null,
                      "range": {
                        "start": 0,
                        "end": 1
                      }
                    },
                    {
                      "type": "Identifier",
                      "name": "setCurrentStep",
                      "typeAnnotation": null,
                      "loc": null,
                      "range": {
                        "start": 0,
                        "end": 1
                      }
                    }
                  ],
                  "loc": null,
                  "range": {
                    "start": 0,
                    "end": 1
                  }
                },
                "init": {
                  "type": "CallExpression",
                  "callee": {
                    "type": "Identifier",
                    "name": "useState",
                    "typeAnnotation": null,
                    "loc": null,
                    "range": {
                      "start": 0,
                      "end": 1
                    }
                  },
                  "arguments": [
                    {
                      "type": "NumericLiteral",
                      "value": 0.0,
                      "loc": null,
                      "range": {
                        "start": 0,
                        "end": 1
                      }
                    }
                  ],
                  "loc": null,
                  "range": {
                    "start": 0,
                    "end": 1
                  }
                },
                "loc": null,
                "range": {
                  "start": 0,
                  "end": 1
                }
              }
            ],
            "loc": null,
            "range": {
              "start": 0,
              "end": 1
            }
          },
          {
            "type": "VariableDeclaration",
            "kind": "const",
            "declarations": [
              {
                "type": "VariableDeclarator",
                "id": {
                  "type": "Identifier",
                  "name": "onSubmit",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 0,
                    "end": 1
                  }
                },
                "init": {
                  "type": "ArrowFunctionExpression",
                  "id": null,
                  "params": [
                    {
                      "type": "Identifier",
                      "name": "errorEvent",
                      "typeAnnotation": null,
                      "loc": null,
                      "range": {
                        "start": 0,
                        "end": 1
                      }
                    }
                  ],
                  "body": {
                    "type": "BlockStatement",
                    "body": [
                      {
                        "type": "ExpressionStatement",
                        "expression": {
                          "type": "CallExpression",
                          "callee": {
                            "type": "Identifier",
                            "name": "logEvent",
                            "typeAnnotation": null,
                            "loc": null,
                            "range": {
                              "start": 0,
                              "end": 1
                            }
                          },
                          "arguments": [
                            {
                              "type": "Identifier",
                              "name": "errorEvent",
                              "typeAnnotation": null,
                              "loc": null,
                              "range": {
                                "start": 0,
                                "end": 1
                              }
                            }
                          ],
                          "loc": null,
                          "range": {
                            "start": 0,
                            "end": 1
                          }
                        },
                        "directive": null,
                        "loc": null,
                        "range": {
                          "start": 0,
                          "end": 1
                        }
                      },
                      {
                        "type": "ExpressionStatement",
                        "expression": {
                          "type": "CallExpression",
                          "callee": {
                            "type": "Identifier",
                            "name": "setCurrentStep",
                            "typeAnnotation": null,
                            "loc": null,
                            "range": {
                              "start": 0,
                              "end": 1
                            }
                          },
                          "arguments": [
                            {
                              "type": "NumericLiteral",
                              "value": 1.0,
                              "loc": null,
                              "range": {
                                "start": 0,
                                "end": 1
                              }
                            }
                          ],
                          "loc": null,
                          "range": {
                            "start": 0,
                            "end": 1
                          }
                        },
                        "directive": null,
                        "loc": null,
                        "range": {
                          "start": 0,
                          "end": 1
                        }
                      }
                    ],
                    "loc": null,
                    "range": {
                      "start": 0,
                      "end": 1
                    }
                  },
                  "generator": false,
                  "async": false,
                  "loc": null,
                  "range": {
                    "start": 0,
                    "end": 1
                  },
                  "expression": false,
                  "loc": null,
                  "range": {
                    "start": 0,
                    "end": 1
                  }
                },
                "loc": null,
                "range": {
                  "start": 0,
                  "end": 1
                }
              }
            ],
            "loc": null,
            "range": {
              "start": 0,
              "end": 1
            }
          },
          {
            "type": "SwitchStatement",
            "discriminant": {
              "type": "Identifier",
              "name": "currentStep",
              "typeAnnotation": null,
              "loc": null,
              "range": {
                "start": 0,
                "end": 1
              }
            },
            "cases": [
              {
                "type": "SwitchCase",
                "test": {
                  "type": "NumericLiteral",
                  "value": 0.0,
                  "loc": null,
                  "range": {
                    "start": 0,
                    "end": 1
                  }
                },
                "consequent": [
                  {
                    "type": "ReturnStatement",
                    "argument": {
                      "type": "JSXElement",
                      "openingElement": {
                        "type": "JSXOpeningElement",
                        "name": {
                          "type": "JSXIdentifier",
                          "name": "OtherComponent",
                          "loc": null,
                          "range": {
                            "start": 0,
                            "end": 1
                          }
                        },
                        "attributes": [
                          {
                            "type": "JSXAttribute",
                            "name": {
                              "type": "JSXIdentifier",
                              "name": "data",
                              "loc": null,
                              "range": {
                                "start": 0,
                                "end": 1
                              }
                            },
                            "value": {
                              "type": "JSXExpressionContainer",
                              "expression": {
                                "type": "ObjectExpression",
                                "properties": [
                                  {
                                    "type": "Property",
                                    "key": {
                                      "type": "Identifier",
                                      "name": "foo",
                                      "typeAnnotation": null,
                                      "loc": null,
                                      "range": {
                                        "start": 0,
                                        "end": 1
                                      }
                                    },
                                    "value": {
                                      "type": "StringLiteral",
                                      "value": "bar",
                                      "loc": null,
                                      "range": {
                                        "start": 0,
                                        "end": 1
                                      }
                                    },
                                    "kind": "init",
                                    "method": false,
                                    "shorthand": false,
                                    "computed": false,
                                    "loc": null,
                                    "range": {
                                      "start": 0,
                                      "end": 1
                                    }
                                  }
                                ],
                                "loc": null,
                                "range": {
                                  "start": 0,
                                  "end": 1
                                }
                              },
                              "loc": null,
                              "range": {
                                "start": 0,
                                "end": 1
                              }
                            },
                            "loc": null,
                            "range": {
                              "start": 0,
                              "end": 1
                            }
                          }
                        ],
                        "selfClosing": true,
                        "loc": null,
                        "range": {
                          "start": 0,
                          "end": 1
                        }
                      },
                      "children": [],
                      "closingElement": null,
                      "loc": null,
                      "range": {
                        "start": 0,
                        "end": 1
                      }
                    },
                    "loc": null,
                    "range": {
                      "start": 0,
                      "end": 1
                    }
                  }
                ],
                "loc": null,
                "range": {
                  "start": 0,
                  "end": 1
                }
              },
              {
                "type": "SwitchCase",
                "test": {
                  "type": "NumericLiteral",
                  "value": 1.0,
                  "loc": null,
                  "range": {
                    "start": 0,
                    "end": 1
                  }
                },
                "consequent": [
                  {
                    "type": "ReturnStatement",
                    "argument": {
                      "type": "JSXElement",
                      "openingElement": {
                        "type": "JSXOpeningElement",
                        "name": {
                          "type": "JSXIdentifier",
                          "name": "OtherComponent",
                          "loc": null,
                          "range": {
                            "start": 0,
                            "end": 1
                          }
                        },
                        "attributes": [
                          {
                            "type": "JSXAttribute",
                            "name": {
                              "type": "JSXIdentifier",
                              "name": "data",
                              "loc": null,
                              "range": {
                                "start": 0,
                                "end": 1
                              }
                            },
                            "value": {
                              "type": "JSXExpressionContainer",
                              "expression": {
                                "type": "ObjectExpression",
                                "properties": [
                                  {
                                    "type": "Property",
                                    "key": {
                                      "type": "Identifier",
                                      "name": "foo",
                                      "typeAnnotation": null,
                                      "loc": null,
                                      "range": {
                                        "start": 0,
                                        "end": 1
                                      }
                                    },
                                    "value": {
                                      "type": "StringLiteral",
                                      "value": "joe",
                                      "loc": null,
                                      "range": {
                                        "start": 0,
                                        "end": 1
                                      }
                                    },
                                    "kind": "init",
                                    "method": false,
                                    "shorthand": false,
                                    "computed": false,
                                    "loc": null,
                                    "range": {
                                      "start": 0,
                                      "end": 1
                                    }
                                  }
                                ],
                                "loc": null,
                                "range": {
                                  "start": 0,
                                  "end": 1
                                }
                              },
                              "loc": null,
                              "range": {
                                "start": 0,
                                "end": 1
                              }
                            },
                            "loc": null,
                            "range": {
                              "start": 0,
                              "end": 1
                            }
                          },
                          {
                            "type": "JSXAttribute",
                            "name": {
                              "type": "JSXIdentifier",
                              "name": "onSubmit",
                              "loc": null,
                              "range": {
                                "start": 0,
                                "end": 1
                              }
                            },
                            "value": {
                              "type": "JSXExpressionContainer",
                              "expression": {
                                "type": "Identifier",
                                "name": "onSubmit",
                                "typeAnnotation": null,
                                "loc": null,
                                "range": {
                                  "start": 0,
                                  "end": 1
                                }
                              },
                              "loc": null,
                              "range": {
                                "start": 0,
                                "end": 1
                              }
                            },
                            "loc": null,
                            "range": {
                              "start": 0,
                              "end": 1
                            }
                          }
                        ],
                        "selfClosing": true,
                        "loc": null,
                        "range": {
                          "start": 0,
                          "end": 1
                        }
                      },
                      "children": [],
                      "closingElement": null,
                      "loc": null,
                      "range": {
                        "start": 0,
                        "end": 1
                      }
                    },
                    "loc": null,
                    "range": {
                      "start": 0,
                      "end": 1
                    }
                  }
                ],
                "loc": null,
                "range": {
                  "start": 0,
                  "end": 1
                }
              },
              {
                "type": "SwitchCase",
                "test": null,
                "consequent": [
                  {
                    "type": "ExpressionStatement",
                    "expression": {
                      "type": "CallExpression",
                      "callee": {
                        "type": "Identifier",
                        "name": "logEvent",
                        "typeAnnotation": null,
                        "loc": null,
                        "range": {
                          "start": 0,
                          "end": 1
                        }
                      },
                      "arguments": [
                        {
                          "type": "StringLiteral",
                          "value": "Invalid step",
                          "loc": null,
                          "range": {
                            "start": 0,
                            "end": 1
                          }
                        }
                      ],
                      "loc": null,
                      "range": {
                        "start": 0,
                        "end": 1
                      }
                    },
                    "directive": null,
                    "loc": null,
                    "range": {
                      "start": 0,
                      "end": 1
                    }
                  },
                  {
                    "type": "ReturnStatement",
                    "argument": {
                      "type": "JSXElement",
                      "openingElement": {
                        "type": "JSXOpeningElement",
                        "name": {
                          "type": "JSXIdentifier",
                          "name": "OtherComponent",
                          "loc": null,
                          "range": {
                            "start": 0,
                            "end": 1
                          }
                        },
                        "attributes": [
                          {
                            "type": "JSXAttribute",
                            "name": {
                              "type": "JSXIdentifier",
                              "name": "data",
                              "loc": null,
                              "range": {
                                "start": 0,
                                "end": 1
                              }
                            },
                            "value": {
                              "type": "JSXExpressionContainer",
                              "expression": {
                                "type": "NullLiteral",
                                "loc": null,
                                "range": {
                                  "start": 0,
                                  "end": 1
                                }
                              },
                              "loc": null,
                              "range": {
                                "start": 0,
                                "end": 1
                              }
                            },
                            "loc": null,
                            "range": {
                              "start": 0,
                              "end": 1
                            }
                          }
                        ],
                        "selfClosing": true,
                        "loc": null,
                        "range": {
                          "start": 0,
                          "end": 1
                        }
                      },
                      "children": [],
                      "closingElement": null,
                      "loc": null,
                      "range": {
                        "start": 0,
                        "end": 1
                      }
                    },
                    "loc": null,
                    "range": {
                      "start": 0,
                      "end": 1
                    }
                  }
                ],
                "loc": null,
                "range": {
                  "start": 0,
                  "end": 1
                }
              }
            ],
            "loc": null,
            "range": {
              "start": 0,
              "end": 1
            }
          }
        ],
        "loc": null,
        "range": {
          "start": 0,
          "end": 1
        }
      },
      "generator": false,
      "async": false,
      "loc": null,
      "range": {
        "start": 0,
        "end": 1
      },
      "loc": null,
      "range": {
        "start": 0,
        "end": 1
      }
    }
  ],
  "sourceType": "script",
  "loc": null,
  "range": {
    "start": 0,
    "end": 1
  }
}
